source/extensions/filters/http/composite/filter.cc

```diff
@@ -0,0 +1,188 @@
+template <class FilterPtrT, class FuncT, class RValT, class... Args>
+RValT delegateFilterAction(FilterPtrT& filter, FuncT func, RValT rval, Args&&... args) {
+  if (filter) {
+    return ((*filter).*func)(std::forward<Args>(args)...);
+  }
+
+  return rval;
+}
+} // namespace
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool end_stream) {
+  decoded_headers_ = true;
+
+  return delegateFilterAction(delegated_filter_, &StreamDecoderFilter::decodeHeaders,
+                              Http::FilterHeadersStatus::Continue, headers, end_stream);
+}
+
+Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_stream) {
+  return delegateFilterAction(delegated_filter_, &StreamDecoderFilter::decodeData,
+                              Http::FilterDataStatus::Continue, data, end_stream);
+}
+
+Http::FilterTrailersStatus Filter::decodeTrailers(Http::RequestTrailerMap& trailers) {
+  return delegateFilterAction(delegated_filter_, &StreamDecoderFilter::decodeTrailers,
+                              Http::FilterTrailersStatus::Continue, trailers);
+}
+
+Http::FilterMetadataStatus Filter::decodeMetadata(Http::MetadataMap& metadata_map) {
+  return delegateFilterAction(delegated_filter_, &StreamDecoderFilter::decodeMetadata,
+                              Http::FilterMetadataStatus::Continue, metadata_map);
+}
+
+void Filter::decodeComplete() {
+  if (delegated_filter_) {
+    delegated_filter_->decodeComplete();
+  }
+}
+
+Http::FilterHeadersStatus Filter::encode100ContinueHeaders(Http::ResponseHeaderMap& headers) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encode100ContinueHeaders,
+                              Http::FilterHeadersStatus::Continue, headers);
+}
+
+Http::FilterHeadersStatus Filter::encodeHeaders(Http::ResponseHeaderMap& headers, bool end_stream) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encodeHeaders,
+                              Http::FilterHeadersStatus::Continue, headers, end_stream);
+}
+Http::FilterDataStatus Filter::encodeData(Buffer::Instance& data, bool end_stream) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encodeData,
+                              Http::FilterDataStatus::Continue, data, end_stream);
+}
+Http::FilterTrailersStatus Filter::encodeTrailers(Http::ResponseTrailerMap& trailers) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encodeTrailers,
+                              Http::FilterTrailersStatus::Continue, trailers);
+}
+Http::FilterMetadataStatus Filter::encodeMetadata(Http::MetadataMap& metadata_map) {
+  return delegateFilterAction(delegated_filter_, &StreamEncoderFilter::encodeMetadata,
+                              Http::FilterMetadataStatus::Continue, metadata_map);
+}
+void Filter::encodeComplete() {
+  if (delegated_filter_) {
+    delegated_filter_->encodeComplete();
+  }
+}
+void Filter::onMatchCallback(const Matcher::Action& action) {
```

Comment:
Coming back to my comments in the docs, my read of this code is that we don't buffer or replay any previous callbacks. So it feels like this can only sanely work right now if matching on request headers? Can we somehow force that at config time?

Reply:
That's the purpose of https://github.com/envoyproxy/envoy/issues/14727, which would support pretty extensive validation of the match tree. I could add a more basic form of validation (match input whitelist) if you think that would be useful

The requirement to match only request headers isn't there for `StreamEncoderFilter`s though, so with the current impl we could match on request + response headers and match. This because the encoder filter wouldn't need to see any callbacks before encodeHeaders. Unfortunately we don't have a great way of validating what kind of filter a filter factory will produce, so right now we can't really make the config validation dependent on whether the requested filter is an encoder/decoder/stream filter. LMK if you even think its valuable to supporting matching on response headers for an encoder filter.

Reply:
One option would be to limit this to request headers initially right now via a whitelist (using the tree validator from #14727 without the full path validation), then introduce a new Action type down the line that allows injecting an encoder filter on the encode path down the line? That way once we get full path validation we can better say "this action can only be done with request header data, while this other action can use both request and response headers".

Reply:
I will defer to you, but my general preference would be to limit this as much as possible, and then we can relax it as we ad more controls? I'm very concerned about people composing things that don't make sense and then having a relatively high support burden? IMO the initial use cases are all going to be request headers so maybe we should start there?

Reply:
Yeah I think that's a good idea, I'll wait for https://github.com/envoyproxy/envoy/pull/15709 to land so we can validate the match tree and remove the encoding paths from this PR

Reply:
This is done